### PR TITLE
Steps toward Hoogle database generation

### PR DESCRIPTION
**Pull request by hdgarrood**
This isn't quite there yet - there are a few unresolved questions I wasn't really sure about:

1. Currently the Hoogle mode isn't exposed via the CLI. Do you think this should become a separate executable, or do you want to incorporate both "modes" into `psc-pages`?

2. The Hoogle mode is aimed towards working on one package at a time - the idea is:

 * you supply a source directory
 * It finds your sources in `src/**/*.purs`, and dependencies in `bower_components/*/src/**/*.purs`
 * The dependencies files are used for desugaring, but not included in the output. (The way I managed  this may be a horrible abuse of the first argument of `P.parseModulesFromFiles`...)
 * Then, the generated file produced only contains declarations from your package.

Might it make sense for the HTML mode to work in a similar way?

### CODE REVIEW

**Comment by hdgarrood** [comment]
Also it's 4am here right now, so I think I should go to sleep and have another look at this when I'm a bit more awake.

File: src/PscPages/AsHoogle.hs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  preamble =
+    T.unlines [ "@package " <> T.pack rpName
+              , "@version " <> T.pack rpVersion
+              , ""
+              ]
+  modules =
+    foldMap renderedModuleAsText rpModules
+
+renderedCodeAsText :: RenderedCode -> Text
+renderedCodeAsText = outputWith elemAsHtml
```
Comment (hdgarrood) [inline src/PscPages/AsHoogle.hs]: Note to self: fix this function name

**Comment by hdgarrood** [comment]
Oh also - an example file that I got from running this on `purescript-maybe`: https://gist.github.com/hdgarrood/af6aea24f19d0365bbed

Also I think we should stop instances from showing up in Hoogle mode. And I think a lot of the functions could be better named.

**Comment by hdgarrood** [comment]
I think this is potentially mergeable now. I'm happy about the way the code is laid out and the way functions are named now. I've tested the executable behaviour (by running once with this code, and once with what's currently in `gh-pages`, and comparing the outputs using `diff`), and I'm pretty certain it is the same. The Hoogle stuff still isn't exposed via the CLI but we can work that out soon.

**Comment by paf31** [comment]
:+1: Thanks!

**Pull request merged**